### ISSUE TITLE
PHP Warning on exec() in include/utils.php

### ISSUE THREAD

**Issue by shokupanjamin** [salesagility/SuiteCRM#9702]
#### Issue
`include/utils.php` throws a PHP Warning at line 588 when `exec()` is disabled by the hosting environment. 

This creates unnecessary logging into the `error_log` file and eventual error log bloat.

#### Expected Behavior
Handle the `exec()` command in a way to avoid throwing a PHP Warning.

#### Actual Behavior
Throws the following PHP Warning into the `error_log` each time `include/utils.php` runs. (One log per minute on my instance):

`PHP Warning: exec() has been disabled for security reasons in {path_to_instance}/include/utils.php on line 588`

Relevant code lines from `/include/utils.php`:
```
function getRunningUser()
{
    // works on Windows and Linux, but might return null on systems that include "exec" in
    // disabled_functions in php.ini (typical in shared hosting)
    $runningUser = exec('whoami');

```

#### Possible Fix
Prefix the `exec()` command with an `@` to avoid error logging.
Alternatively, try to detect whether `exec()` is disabled in the environment before attempting to run it.

#### Steps to Reproduce
Runs automatically - I'm assuming it is triggered by the cron.

#### Your Environment
* SuiteCRM Version used: 7.12.6
* Browser name and version: Issue seems to be a server-side cron task, so browser is not relevant.
* Environment name and version (e.g. MySQL, PHP 7): PHP 7.4
* Operating System and version (e.g Ubuntu 16.04): Apache

Unfortunately, my hosting config doesn't allow me to disable logging of PHP Warnings without disabling PHP error logging entirely. :(

**Comment by chris001** [comment]
Good catch.  It was unnecessary to call `exec("whoami")` which requires special privileges to make a command line call, like that one, and therefore that line fails on nearly all shared hosting servers as it should, when instead you should call `get_current_user()` which requires no privileges.and works fine on all shared hosting servers.

**Comment by pgorod** [comment]
In this documentation page https://www.php.net/manual/en/function.get-current-user.php

There is a comment saying: 
> On Linux this function is indeed returning the owner of the script. If you want to know the username PHP is running as you can use POSIX functions (or shell_exec with 'whoami').

If that is correct (and it might not be, it's just some person commenting) then **get_current_user** is not what we want.

I did extensive (and I mean extensive) checking and testing of all possibilities when I wrote that code. I tried it in Linux, in Windows, and in several shared hostings. As you can see from my comments there, the possibility that **exec** wouldn't be available was taken in consideration; only at the time, it didn't throw a warning, now it does with newer PHPs.

I think the solutions @shokupanjamin describes in his "Possible Fix" section would be fine here.

Other solutions changing the functions called, all I can say is: sure, if you test them thoroughly in several environments, which is a lot of work...

**Comment by chris001** [comment]
```
  $processUserData = posix_getpwuid(posix_geteuid());
  $runningUser = $processUserData['name'];
```
This works on all hosting @shokupanjamin

**Comment by pgorod** [comment]
Well, I don't suppose you tried it on _all_hostings... I read somewhere that some hostings disable posix functions for "security" reasons. You can find a few such situations in Google.

But your fix is good, it leaves the code testing if the posix functions are available or not, so 🆒

**Comment by chris001** [comment]
To test on all hosting, how about run tests on top 10 Linux distros, Windows, MacOS.  Github or Travis can do this.

**Comment by pgorod** [comment]
But that won't tell you how some hosting company somewhere decided to configure their servers... Which is the real problem with this particular bit of code...

**Comment by chris001** [comment]
You can configure one of the test runs as a high security PHP test server with the list of highest security PHP functions disabled, to find out if it detects and informs the user without log bloat or malfunction.

**Comment by gunnicom** [comment]
Maybe you can check for disabled functions first like:

`function exec_enabled() {
  $disabled = explode(',', ini_get('disable_functions'));
  return !in_array('exec', $disabled);
}`

**Comment by chris001** [comment]
Example high secure PHP disabled functions to run tests on:
`disable_functions=pcntl_alarm,pcntl_fork,pcntl_waitpid,pcntl_wait,pcntl_wifexited,pcntl_wifstopped,pcntl_wifsignaled,pcntl_wexitstatus,pcntl_wtermsig,pcntl_wstopsig,pcntl_signal,pcntl_signal_dispatch,pcntl_get_last_error,pcntl_strerror,pcntl_sigprocmask,pcntl_sigwaitinfo,pcntl_sigtimedwait,pcntl_exec,pcntl_getpriority,pcntl_setpriority,system, exec, passthru, shell, shell_exec, popen, pclose, proc_nice, proc_terminate, proc_get_status, proc_close, pfsockopen, leak, apache_child_terminate, posix_kill, posix_mkfifo, posix_setpgid, posix_setsid, posix_setuid, escapeshellcmd, escapeshellarg, phpinfo, proc_open, show_source, passthru`

**Comment by chris001** [comment]
Set 2:
`disable_function=apache_child_terminate, apache_get_modules, apache_note, apache_setenv, define_syslog_variables, disk_free_space, disk_total_space, diskfreespace, dl, escapeshellarg, escapeshellcmd, exec, extract, get_cfg_var, get_current_user, getcwd, getenv, getlastmod, getmygid, getmyinode, getmypid, getmyuid, ini_restore, ini_set, passthru, pcntl_alarm, pcntl_exec, pcntl_fork, pcntl_get_last_error, pcntl_getpriority, pcntl_setpriority, pcntl_signal, pcntl_signal_dispatch, pcntl_sigprocmask, pcntl_sigtimedwait, pcntl_sigwaitinfo, pcntl_strerrorp, pcntl_wait, pcntl_waitpid, pcntl_wexitstatus, pcntl_wifexited, pcntl_wifsignaled, pcntl_wifstopped, pcntl_wstopsig, pcntl_wtermsig, php_uname, phpinfo, popen, posix_getlogin, posix_getpwuid, posix_kill, posix_mkfifo, posix_setpgid, posix_setsid, posix_setuid, posix_ttyname, posix_uname, posixc, proc_close, proc_get_status, proc_nice, proc_open, proc_terminate, ps_aux, putenv, readlink, runkit_function_rename, shell_exec, show_source, symlink, syslog, system`

**Comment by pgorod** [comment]
The code already has the necessary check for the function it tries to use

https://github.com/salesagility/SuiteCRM/blob/master/include/utils.php#L593

All it needs is the improved warning-suppression and error-handling that is in the proposed fix
